tensorflow_similarity/losses/arcface_loss.py

```diff
@@ -0,0 +1,232 @@
+    """Implement of ArcFace: Additive Angular Margin Loss:
+            Step 1: Create a trainable kernel matrix with the shape of [embedding_size, num_classes].
+            Step 2: Normalize the kernel and prediction vectors.
+            Step 3: Calculate the cosine similarity between the normalized prediction vector and the kernel.
+            Step 4: Create a one-hot vector include the margin value for the ground truth class.
+            Step 5: Add margin_hot to the cosine similarity and multiply it by scale.
+            Step 6: Calculate the cross-entropy loss.
+
+            ArcFace: Additive Angular Margin Loss for Deep Face
+                     Recognition. [online] arXiv.org. Available at:
+                     <https://arxiv.org/abs/1801.07698v3>.
+
+            Standalone usage:
+                >>> loss_fn = tfsim.losses.ArcFaceLoss(num_classes=2, embedding_size=3)
+                >>> labels = tf.Variable([1, 0])
+                >>> embeddings = tf.Variable([[0.2, 0.3, 0.1], [0.4, 0.5, 0.5]])
+                >>> loss = loss_fn(labels, embeddings)
+            Args:
+                num_classes: Number of classes.
+                embedding_size: The size of the embedding vectors.
+                margin: The margin value.
+                scale: s in the paper, feature scale
+                name: Optional name for the operation.
+                reduction: Type of loss reduction to apply to the loss.
+            """
+
+    def __init__(self,
+                 num_classes: int,
+                 embedding_size: int,
+                 margin: float = 0.50,  # margin in radians
+                 scale: float = 64.0,  # feature scale
+                 name: Optional[str] = "ArcFaceLoss",
+                 reduction: Callable = tf.keras.losses.Reduction.AUTO,
+                 **kwargs: object) -> object:
+        super().__init__(reduction=reduction, name=name, **kwargs)
+
+        self.num_classes = num_classes
+        self.embedding_size = embedding_size
+        self.margin = margin
+        self.scale = scale
+        self.name = name
+        self.kernel = tf.Variable(tf.random.normal([embedding_size, num_classes]))
+
+    def call(self, y_true: FloatTensor, y_pred: FloatTensor) -> FloatTensor:
+        y_pred_norm = tf.math.l2_normalize(y_pred, axis=1)
+        kernel_norm = tf.math.l2_normalize(self.kernel, axis=0)
+
+        cos_theta = tf.matmul(y_pred_norm, kernel_norm)
+        cos_theta = tf.clip_by_value(cos_theta, -1.0, 1.0)
+
+        m_hot = tf.one_hot(y_true, self.num_classes, on_value=self.margin, axis=1)
+        m_hot = tf.reshape(m_hot, [-1, self.num_classes])
+
+        cos_theta = tf.acos(cos_theta)
+        cos_theta += m_hot
+        cos_theta = tf.math.cos(cos_theta)
+        cos_theta = tf.math.multiply(cos_theta, self.scale)
+
+        cce = tf.keras.losses.SparseCategoricalCrossentropy(from_logits=True, reduction=self.reduction)
+        loss: FloatTensor = cce(y_true, cos_theta)

```

Comment:
At the same time, I change the type of the loss as FloatTensor, such that the other issue need to be solved as well.